include/Dialects/LinalgTransform/TransformOpInterface.h

```diff
@@ -0,0 +1,226 @@
+                                   FnTy transform) {
+  using TransformOpType =
+      typename llvm::function_traits<FnTy>::template arg_t<0>;
+  for (Operation *target : targets) {
+    auto specificOp =
+        functional::detail::IsaOr<TransformOpType>::dyn_cast(target);
+    if (!specificOp)
+      return failure();
+
+    auto result = transform(specificOp);
+    if (failed(result))
+      return failure();
+
+    results.push_back(*result);
+  }
+  return success();
+}
+
+/// Trait implementing the TransformOpInterface for operations applying a
+/// transformation to a single operation handle and producing a single operation
+/// handle. The op must implement:
+///   - FailureOr<convertible-to-Operation*> applyToOne(OpTy) - a transformation
+///   that is applied in turn to all payload IR operations that correspond to
+///   the handle of the transform IR operation.
+/// In the functions above, OpTy is either Operation * or a concrete payload IR
+/// Op class that the transformation is applied to (NOT the class of the
+/// transform IR op).
+template <typename OpTy>
+class TargetableOpTrait : public OpTrait::TraitBase<OpTy, TargetableOpTrait> {
+public:
+  /// Applies the transformation to each op from the only target and sets the
+  /// only result to correspond to the list of individual results.
+  LogicalResult apply(TransformResults &transformResults,
+                      TransformState &state) {
+    using TransformOpType = typename llvm::function_traits<
+        decltype(&OpTy::applyToOne)>::template arg_t<0>;
+    ArrayRef<Operation *> targets =
+        state.getTargets(this->getOperation()->getOperand(0));
+    SmallVector<Operation *> results;
+    if (failed(applyTransformToEach(
+            targets, results, [&](TransformOpType specificOp) {
+              return static_cast<OpTy *>(this)->applyToOne(specificOp);
+            })))
+      return failure();
+    transformResults.set(0, results);
+    return success();
+  }
+
+  /// Verifies that the op satisfies the requirements for this trait.
+  static LogicalResult verifyTrait(Operation *) {
+    static_assert(OpTy::template hasTrait<OpTrait::OneOperand>(),
+                  "expected single-operand op");
+    static_assert(OpTy::template hasTrait<OpTrait::OneResult>(),
+                  "expected single-result op");
+    return success();
+  }
+};
+
+/// Trait implementing the TransformOpInterface for operations applying a
+/// transformation to a single operation handle and not producing any handles.
+/// The op must implement:
+///   - LogicalResult applyToOne(OpTy) - a transformation that is applied in
+///   turn to all payload IR operations that correspond to the handle of the
+///   transform IR operation.
+/// In the functions above, OpTy is either Operation * or a concrete payload IR
+/// Op class that the transformation is applied to (NOT the class of the
+/// transform IR op).
+template <typename OpTy>
+class TargetableNoResultOpTrait
```

Comment:
nit: should `SingleOperandZeroResult` or something similar appear in the trait name ?